lib/vmpooler/pool_manager.rb

```diff
@@ -458,6 +461,77 @@ def _check_snapshot_queue
+      datacenter = pool['datacenter'] if pool.key?('datacenter')
+      datacenter = default_datacenter if datacenter.nil?
+      return if cluster.nil?
+      return if datacenter.nil?
+      { 'cluster' => cluster, 'datacenter' => datacenter }
+    end
+
+    def select_hosts(pool_name, provider, provider_name, cluster, datacenter, percentage)
+      $provider_hosts[provider_name] = {} unless $provider_hosts.key?(provider_name)
+      $provider_hosts[provider_name][datacenter] = {} unless $provider_hosts[provider_name].key?(datacenter)
+      $provider_hosts[provider_name][datacenter][cluster] = {} unless $provider_hosts[provider_name][datacenter].key?(cluster)
+      $provider_hosts[provider_name][datacenter][cluster]['checking'] = true
+      hosts_hash = provider.select_target_hosts(cluster, datacenter, percentage)
+      $provider_hosts[provider_name][datacenter][cluster] = hosts_hash
+      $provider_hosts[provider_name][datacenter][cluster]['check_time_finished'] = Time.now
+    end
+
+    def run_select_hosts(provider, pool_name, provider_name, cluster, datacenter, max_age, percentage)
+      now = Time.now
+      if $provider_hosts.key?(provider_name) and $provider_hosts[provider_name].key?(datacenter) and $provider_hosts[provider_name][datacenter].key?(cluster) and $provider_hosts[provider_name][datacenter][cluster].key?('checking')
+        wait_for_host_selection(pool_name, provider_name, cluster, datacenter)
+      elsif $provider_hosts.key?(provider_name) and $provider_hosts[provider_name].key?(datacenter) and $provider_hosts[provider_name][datacenter].key?(cluster) and $provider_hosts[provider_name][datacenter][cluster].key?('check_time_finished')
+        select_hosts(pool_name, provider, provider_name, cluster, datacenter, percentage) if now - $provider_hosts[provider_name][datacenter][cluster]['check_time_finished'] > max_age
+      else
+        select_hosts(pool_name, provider, provider_name, cluster, datacenter, percentage)
+      end
+    end
+
+    def wait_for_host_selection(pool_name, provider_name, cluster, datacenter, maxloop = 0, loop_delay = 5, max_age = 60)
+      loop_count = 1
+      until $provider_hosts[provider_name][datacenter][cluster].key?('check_time_finished')
+        sleep(loop_delay)
+        unless maxloop.zero?
+          break if loop_count >= maxloop
+          loop_count += 1
+        end
+      end
+      return unless $provider_hosts[provider_name][datacenter][cluster].key?('check_time_finished')
+      loop_count = 1
+      while Time.now - $provider_hosts[provider_name][datacenter][cluster]['check_time_finished'] > max_age
+        sleep(loop_delay)
+        unless maxloop.zero?
+          break if loop_count >= maxloop
+          loop_count += 1
+        end
+      end
+    end
+
+    def select_next_host(provider_name, datacenter, cluster, architecture)
+      provider_hosts = $provider_hosts
+      host = provider_hosts[provider_name][datacenter][cluster]['architectures'][architecture][0]
```

Comment:
I think if `provider_hosts[provider_name][datacenter][cluster]['architectures'][architecture]` is a string then `[0]` could return the first character which is not what you probably expect here?

Reply:
That's true. It's populated by select_least_used_hosts from lib/vmpooler/providers/vsphere.rb:607 which returns an array. Do you think I should be more explicit than this at some point to ensure an array is returned?

Reply:
Ok I got it, thanks